File: 2015/1/elixir/main.exs
defmodule Main do
  {_, input} =
    File.read("../input.txt")

  input = input |> String.trim() |> String.graphemes()

  input
  |> Enum.reduce(0, fn step, floor ->
    case step do
      "(" -> floor + 1
      ")" -> floor - 1
    end
  end)
  |> IO.puts()

  input
  |> Enum.reduce_while({0, 0}, fn step, i_floor ->
    {i, floor} = i_floor

    i = i + 1

    floor =
      case step do
        "(" -> 1
        ")" -> -1
      end + floor

    case floor do
      -1 -> {:halt, {i, floor}}
      _ -> {:cont, {i, floor}}
    end
  end)
  |> elem(0)
  |> IO.puts()
end

File: 2015/1/rust/src/main.rs
use std::{fs, ops::ControlFlow};

fn main() {
    let input: String = match fs::read_to_string("../input.txt") {
        Ok(input) => input,
        _ => panic!("invalid input!!!"),
    };
    let input = input.trim();

    let part1 = input.chars().fold(0, |floor, v| {
        (match v {
            '(' => 1,
            ')' => -1,
            _ => 0,
        }) + floor
    });

    println!("{}", part1.to_string());

    let part2 = input.chars().try_fold((0, 1), |state, v| {
        let (floor, step) = state;
        let floor = (match v {
            '(' => 1,
            ')' => -1,
            _ => 0,
        }) + floor;
        match floor {
            -1 => ControlFlow::Break(step),
            _ => ControlFlow::Continue((floor, step + 1)),
        }
    });
    let part2 = match part2 {
        ControlFlow::Break(part2) => part2,
        _ => panic!("something bad happened"),
    };

    println!("{}", part2.to_string());
}
